### PR TITLE
[TextInputLayout] Speedup simple inflate

### LINKED ISSUE #1745: [TextInputLayout] TextInputLayout with TextInputEditText drastically slowing down inflation.
**Description:** When inflating a layout, the presence of a TextInputLayout makes the inflation noticably slower. It's enough to add two TILs to a layout to start noticing a lag on my device (nokia 7.1, which is not high-end but not the worst phone either). It happens even when I put them in a simple linearlayout with no other views. On the other hand, if I inflate a layout with even 10 EditTexts, there is no lag whatsoever.

You can reproduce it on this basic sample (the dashboard fragment contains a few TILs and the notifications one contains normal EditTexts): https://drive.google.com/file/d/1DwvxzHM74TChIJGuttkRF6SRPcb8fEGd/view?usp=sharing

Here is a screenrecord using the sample I provided. First I navigate to the fragment with basic EditTexts and there is no lag, then I navigate to the one with TILs and there's the lag. It might not look that severe, but in a layout with other widgets the inflation times add up and it's enough to have a single TIL to get a bad UX.
https://drive.google.com/file/d/12P-XErDnZjSuhIwSqV6f9ydzYHQGUDZU/view?usp=sharing 

previously filed at https://issuetracker.google.com/issues/168869410 , was told to refile here

**Expected behavior:** TextInputLayout should inflate much faster, at least so that it doesn't cause any lags in fragment transitions when you have a few of them in your layout

**Source code:** https://drive.google.com/file/d/1DwvxzHM74TChIJGuttkRF6SRPcb8fEGd/view?usp=sharing

**Android API version:** Tested with API 29

**Material Library version:** 1.3.0-alpha02

**Device:** reproduced on Nokia 7.1 with Android 10 (physical device), and a virtual Pixel 3 XL with A10

### LINKED ISSUE #2160: [TextInputLayout] TextInputLayout with TextInputEditText drastically slowing down inflation
Description: When inflating a layout, the presence of a TextInputLayout makes the inflation noticably slower. It's enough to add two TILs to a layout to start noticing a lag on my device (nokia 7.1, which is not high-end but not the worst phone either). It happens even when I put them in a simple linearlayout with no other views. On the other hand, if I inflate a layout with even 10 EditTexts, there is no lag whatsoever.

You can reproduce it on this basic sample (the dashboard fragment contains a few TILs and the notifications one contains normal EditTexts): https://drive.google.com/file/d/1DwvxzHM74TChIJGuttkRF6SRPcb8fEGd/view?usp=sharing

Here is a screenrecord using the sample I provided. First I navigate to the fragment with basic EditTexts and there is no lag, then I navigate to the one with TILs and there's the lag. It might not look that severe, but in a layout with other widgets the inflation times add up and it's enough to have a single TIL to get a bad UX.
https://drive.google.com/file/d/12P-XErDnZjSuhIwSqV6f9ydzYHQGUDZU/view?usp=sharing

previously filed at https://issuetracker.google.com/issues/168869410 , was told to refile here

Expected behavior: TextInputLayout should inflate much faster, at least so that it doesn't cause any lags in fragment transitions when you have a few of them in your layout

Source code: https://drive.google.com/file/d/1DwvxzHM74TChIJGuttkRF6SRPcb8fEGd/view?usp=sharing

Android API version: Tested with API 29

Material Library version: 1.3.0-alpha02

### PR DESCRIPTION
**Pull request by slavonnet**
1. By default, create all children as Gone until the first state of the Visible. This greatly reduces nesting and shortens initialization and child recalculation times. The changes are minimal and after the first setVisible on the child, it results in a pre-PR state on the parent.
2. The order has been changed. First setVisibility (GONE) and then addView () on endIconView. Trace showed long inflates at CheckableImageButton.

Significant speedup requires permanent hiding of the tree if all children are hidden or if the element is off-screen, thereby significantly shortening the tree for calculation. For example, the NavigationView full tree can be hidden (and removed from recalculation) until onDrug or the panel is shown.

closes #1745
closes #2160

### CODE REVIEW

**Comment by slavonnet** [comment]
@drchen please RV

**Comment by yamin8000** [comment]
@slavonnet you wrote in English yet I can't understand you correctly. What exactly is your solution? can you show it with code?

**Comment by slavonnet** [comment]
@yamin8000 What exactly is not possible to understand? My fix fixes the problem of long inflation with a large number of elements by decreasing the tree depth.

**Comment by yamin8000** [comment]
> 
> 
> @yamin8000 What exactly is not possible to understand? My fix fixes the problem of long inflation with a large number of elements by decreasing the tree depth. It also excludes from the calculation of the Image Button, which takes a long time, but is not displayed.

My bad, I didn't saw your commits first time.
Yeah about that `CheckableImageButton`, I saw long inflation of that in my profiler Trace and I don't even know why is that related to `TextInputLayout`.

File: lib/java/com/google/android/material/textfield/TextInputLayout.java [hunk 1]
```diff
@@ -2521,10 +2529,54 @@ private void updateSuffixTextVisibility() {
...
     }
     updateDummyDrawables();
+    updateEndLayoutVisible();
+  }
+
+  @SuppressWarnings("ConstantConditions")
+  private void updateEndLayoutVisible() {
+    if (endIconView == null || endIconFrame == null || errorIconView == null
+        || suffixTextView == null || endLayout == null)
+      return;;
```
Comment (drchen) [inline lib/java/com/google/android/material/textfield/TextInputLayout.java:2544]: Please always add {}, rest the same.

File: lib/java/com/google/android/material/textfield/TextInputLayout.java [hunk 2]
```diff
@@ -2521,10 +2529,54 @@ private void updateSuffixTextVisibility() {
...
+
+    if (!hasEndIcon() && endIconView.getVisibility() != GONE)
+      endIconView.setVisibility(GONE);
+
+    if (endIconFrame.getVisibility() != endIconView.getVisibility())
+      endIconFrame.setVisibility(endIconView.getVisibility());
+
+    int endIconFrameVis = endIconFrame.getVisibility();
```
Comment (drchen) [inline lib/java/com/google/android/material/textfield/TextInputLayout.java:2552]: Don't use uncommon abbreviations, rest the same.

File: lib/java/com/google/android/material/textfield/TextInputLayout.java [hunk 3]
```diff
@@ -2521,10 +2529,54 @@ private void updateSuffixTextVisibility() {
...
     suffixTextView.setVisibility(visible ? VISIBLE : GONE);
+    updateEndLayoutVisible();
     if (oldSuffixVisibility != suffixTextView.getVisibility()) {
       getEndIconDelegate().onSuffixVisibilityChanged(visible);
     }
     updateDummyDrawables();
+    updateEndLayoutVisible();
+  }
+
+  @SuppressWarnings("ConstantConditions")
```
Comment (drchen) [inline lib/java/com/google/android/material/textfield/TextInputLayout.java:2540]: Q) Why do we need to suppress the warning here and there?

**Review by drchen — COMMENTED** [review]
I'll leave this to @leticiarossi for a more thorough review. Just left a few basic comments.

File: lib/java/com/google/android/material/textfield/TextInputLayout.java (hunk 2 above)
Comment (slavonnet) [reply lib/java/com/google/android/material/textfield/TextInputLayout.java:2573]: Give example

**Comment by slavonnet** [comment]
PR look strange after sync master...

**Comment by Berki2021** [comment]
I am looking forward to this update. TextInputLayout destroys my app-performance...

**Comment by slavonnet** [comment]
> I am looking forward to this update. TextInputLayout destroys my app-performance...

[1.5.0-alpha999.zip](https://github.com/material-components/material-components-android/files/7020319/1.5.0-alpha999.zip)

File: lib/java/com/google/android/material/textfield/TextInputLayout.java (hunk 2 above)
Comment (drchen) [reply lib/java/com/google/android/material/textfield/TextInputLayout.java:2573]: Please replace vis with visibility

File: lib/java/com/google/android/material/textfield/TextInputLayout.java [hunk 4]
```diff
@@ -2541,13 +2556,62 @@ public TextView getSuffixTextView() {
...
+    }
+  }
+
+  private void updateStartLayoutVisible() {
+    final int startIconVisibility = startIconView.getVisibility();
+    final int prefixVisibility = prefixTextView.getVisibility();
+    final int startLayoutVisibility = startLayout.getVisibility();
+
+    if (startIconVisibility == VISIBLE || prefixVisibility == VISIBLE) {
+      if (startLayoutVisibility != VISIBLE) {
```
Comment (drchen) [inline lib/java/com/google/android/material/textfield/TextInputLayout.java:2607]: nit) Not needed. I think the View class have the check. Rest the same.

File: lib/java/com/google/android/material/textfield/TextInputLayout.java [hunk 5]
```diff
@@ -2541,13 +2556,62 @@ public TextView getSuffixTextView() {
...
+      }
+    }
+  }
+
+  private void updateStartLayoutVisible() {
+    final int startIconVisibility = startIconView.getVisibility();
+    final int prefixVisibility = prefixTextView.getVisibility();
+    final int startLayoutVisibility = startLayout.getVisibility();
+
+    if (startIconVisibility == VISIBLE || prefixVisibility == VISIBLE) {
```
Comment (drchen) [inline lib/java/com/google/android/material/textfield/TextInputLayout.java:2606]: Can just simplify this as: 

startLayout. setVisibility(startIconVisibility == VISIBLE || prefixVisibility == VISIBLE ? VISIBLE : GONE);

Rest the same.

File: lib/java/com/google/android/material/textfield/TextInputLayout.java [hunk 6]
```diff
@@ -4287,11 +4355,18 @@ private void updateStrokeErrorColor(boolean hasFocus, boolean isHovered) {
 
   private void setErrorIconVisible(boolean errorIconVisible) {
     errorIconView.setVisibility(errorIconVisible ? VISIBLE : GONE);
-    endIconFrame.setVisibility(errorIconVisible ? GONE : VISIBLE);
+    if (errorIconVisible && isEndIconVisible())
```
Comment (drchen) [inline lib/java/com/google/android/material/textfield/TextInputLayout.java:4358]: Always add {}

**Comment by leticiarossi** [comment]
I had to make a few modifications from your original change but hopefully there's an improvement now in the performance time

**Pull request closed without merging**